cmd/controller-manager/main.go

```diff
@@ -141,8 +141,23 @@ func main() {
 	bsController := backupschedule.NewController(kubeCli, cli, informerFactory, kubeInformerFactory)
 	controllerCtx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	go informerFactory.Start(controllerCtx.Done())
-	go kubeInformerFactory.Start(controllerCtx.Done())
+
+	// Start informer factories after all controller are initialized.
+	informerFactory.Start(controllerCtx.Done())
+	kubeInformerFactory.Start(controllerCtx.Done())
+
+	// Wait for all started informers' cache were synced.
+	for v, synced := range informerFactory.WaitForCacheSync(wait.NeverStop) {
+		if !synced {
+			glog.Fatalf("error syncing informer for %v", v)
```

Comment:
When does this happen?

Reply:
e.g. apiserver errors, the controller manager should restart to reconnect again

Reply:
Okay, sgtm.

Reply:
This logic can only be triggered during tidb-controller-manager start, if it starts successfully, the restart will happen due to leader lost with kube-apiserver errors, right?

Reply:
yes, and of course other fatal errors or panics in the code.